pennylane/beta/tapes/tape.py

```diff
@@ -833,3 +835,105 @@ def data(self):
     @data.setter
     def data(self, params):
         self.set_parameters(params, trainable_only=False)
+
+    # ========================================================
+    # execution methods
+    # ========================================================
+
+    def execute(self, device, params=None):
+        """Execute the tape on a quantum device.
+
+        Args:
+            device (~.Device): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameters are used (via :meth:`~.get_parameters`).
+
+        **Example**
+
+        .. code-block:: python
+
+            from pennylane.beta.tapes import QuantumTape
+            from pennylane.beta.queuing import expval, var, sample, probs
+
+            with QuantumTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                probs(wires=[0, 'a'])
+
+        If parameters are not provided, the existing tape parameters are used:
+
+        >>> dev = qml.device("default.qubit", wires=[0, 'a'])
+        >>> tape.execute(dev)
+        array([[8.84828969e-01, 3.92449987e-03, 4.91235209e-04, 1.10755296e-01]])
+
+        Parameters can be optionally passed during execution:
+
+        >>> tape.execute(dev, params=[1.0, 0.0, 1.0])
+        array([[0.5931328 , 0.17701835, 0.05283049, 0.17701835]])
+
+        Parameters provided for execution are temporary, and do not affect
+        the tapes' parameters in-place:
+
+        >>> tape.get_parameters()
+        [0.432, 0.543, 0.133]
+        """
+        if params is None:
+            params = self.get_parameters()
+
+        return self._execute(params, device=device)
+
+    def execute_device(self, params, device):
+        """Execute the tape on a quantum device.
+
+        This is a low-level method, intended to be called by an interface,
+        and does not support autodifferentiation.
+
+        For more details on differentiable tape execution, see :meth:`~.execute`.
+
+        Args:
+            device (~.Device): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+        """
+        device.reset()
```

Comment:
will this introduce overhead for every execution? is it unavoidable?

Reply:
That's a good question.

With the way devices currently work, I _think_ this in unavoidable (this is how the current QNode execution + device API works, so I didn't want to make a logical change here). But this is worth exploring later on, especially in the context of caching the pre-rotated state and re-using it.